--- init.orig	2022-07-24 03:20:49.045113593 +1000
+++ init	2022-07-24 03:22:43.509121318 +1000
@@ -393,8 +393,12 @@
       esac
     fi
   elif [ "$1" != 'k' ];then
-    SPEC_VAR="${1}drv" # defined in boot parameter
+    SPEC_VAR="DRV_${1}" # defined in DRV_SPECS file
     eval ONE_SFS_SPEC="\$$SPEC_VAR"
+    if [ "$ONE_SFS_SPEC" = "" ];then
+      SPEC_VAR="${1}drv" # defined in boot parameter
+      eval ONE_SFS_SPEC="\$$SPEC_VAR"
+    fi
   fi
   if [ "$ONE_SFS_SPEC" = "" ];then
     if [ "$ROOTFS_SPEC" ];then
@@ -962,6 +966,19 @@
 P_PART="$SPEC_PART"
 P_MP="$SPEC_MP"
 
+# process DRV_SPECS in install directory, if present
+DRV_SPEC_FFN="${P_MP}${PSUBDIR}/DRV_SPECS"
+if [ -f "$DRV_SPEC_FFN" ]; then
+  echo "Using $DRV_SPEC_FFN"
+  # ensure only DRV variables get used
+  SPEC_TMP_FN="/tmp/${DRV_SPEC_FFN##*/}.txt"
+  tr -d '\r' < "$DRV_SPEC_FFN" | grep "^DRV_" > "$SPEC_TMP_FN"
+  . "$SPEC_TMP_FN"
+  [ "$DRV_AP" ] && APPEND_DRVS="$DRV_AP"
+  [ "$DRV_PRE" ] && PREPEND_DRVS="$DRV_PRE"
+  rm "$SPEC_TMP_FN"
+fi
+
 #have basic system, now try to add optional stuff
 DRVS_TMP_FN='/tmp/drvs_defs.txt'
 [ -e "$DRVS_TMP_FN" ] && rm "$DRVS_TMP_FN"
