--- init.orig	2022-06-14 12:58:28.272090526 +1000
+++ init	2022-06-14 13:01:11.124101517 +1000
@@ -948,6 +948,19 @@
 P_PART="$SPEC_PART"
 P_MP="$SPEC_MP"
 
+# process DRV_SPECS in install directory, if present
+DRV_SPEC_FFN="${P_MP}${PSUBDIR}/DRV_SPECS"
+if [ -f "$DRV_SPEC_FFN" ]; then
+  echo "Using $DRV_SPEC_FFN"
+  # ensure only DRV variables get used
+  SPEC_TMP_FN="/tmp/${DRV_SPEC_FFN##*/}.txt"
+  tr -d '\r' < "$DRV_SPEC_FFN" | grep "^DRV_" > "$SPEC_TMP_FN"
+  . "$SPEC_TMP_FN"
+  [ "$DRV_AP" ] && APPEND_DRVS="$DRV_AP"
+  [ "$DRV_PRE" ] && PREPEND_DRVS="$DRV_PRE"
+  rm "$SPEC_TMP_FN"
+fi
+
 #have basic system, now try to add optional stuff
 DRVS_TMP_FN='/tmp/drvs_defs.txt'
 [ -e "$DRVS_TMP_FN" ] && rm "$DRVS_TMP_FN"
