--- init.git	2022-07-20 21:45:08.919896949 +1000
+++ init-woof-patched	2022-07-24 18:06:38.446004822 +1000
@@ -297,10 +297,10 @@
   if [ "$ONE_MP" != "$SAVE_MP" ];then
    #ensure SAVEPART is mounted on /mnt/dev_save
    [ -d "$SAVE_MP" ] || mkdir -p $SAVE_MP
-   OPT="move"
-   [ "$UNIONFS" = 'overlay' ] && OPT="bind"
-   echo "mount -o $OPT $ONE_MP $SAVE_MP" #debug
-   mount -o $OPT $ONE_MP $SAVE_MP
+   echo "mount -o move $ONE_MP $SAVE_MP" #debug
+   mount -o move $ONE_MP $SAVE_MP
+   #re-establish P_MP if PDRV was the moved partition MP
+   [ "$SAVEPART" = "$P_PART" ] && P_MP="$SAVE_MP"
   fi
   SAVE_FS="$ONE_FS"
  else
