--- rc.shutdown.orig	2022-09-25 11:37:00.613004000 +1000
+++ rc.shutdown	2022-09-28 11:13:32.475014138 +1000
@@ -399,6 +399,14 @@
 fi
 
 # remount PUNIONFS read-only
+if [ "$PUNIONFS" = 'overlay' ]; then
+	busybox mount -no remount,ro / > /dev/console 2>&1
+	if [ "$SAVE_LAYER" ];then
+		SAVEMP="$(stat -Lc %m "/initrd${SAVE_LAYER}")"
+		SAVEDEV=`grep "$SAVEMP" /proc/mounts | cut -f 1 -d ' '`
+		[ "$SAVEDEV" ] && busybox mount -no remount,ro "$SAVEMP" > /dev/console 2>&1
+	fi
+fi
 if [ "$SAVE_LAYER" ];then
 	SAVEDEV=`grep "/initrd${SAVE_LAYER}" /proc/mounts | cut -f 1 -d ' '`
 	SAVEFS=`grep "/initrd${SAVE_LAYER}" /proc/mounts | cut -f 3 -d ' '`
